zerver/actions/user_status.py

```diff
@@ -35,21 +40,58 @@ def do_update_user_status(
         emoji_name=emoji_name,
         emoji_code=emoji_code,
         reaction_type=reaction_type,
+        scheduled_end_time=scheduled_end_time,
     )
 
-    event = dict(
-        type="user_status",
-        user_id=user_profile.id,
-    )
-
+    updated_status_fields: dict[str, Any] = {}
     if away is not None:
-        event["away"] = away
+        updated_status_fields["away"] = away
 
     if status_text is not None:
-        event["status_text"] = status_text
+        updated_status_fields["status_text"] = status_text
 
     if emoji_name is not None:
-        event["emoji_name"] = emoji_name
-        event["emoji_code"] = emoji_code
-        event["reaction_type"] = reaction_type
-    send_event_on_commit(realm, event, get_user_ids_who_can_access_user(user_profile))
+        updated_status_fields["emoji_name"] = emoji_name
+        updated_status_fields["emoji_code"] = emoji_code
+        updated_status_fields["reaction_type"] = reaction_type
+
+    user_ids_to_send_event = get_user_ids_who_can_access_user(user_profile)
+    if not isinstance(scheduled_end_time, MissingType):
+        event = dict(
+            type="user_status",
+            user_id=user_profile.id,
+            scheduled_end_time=scheduled_end_time,
+            **updated_status_fields,
+        )
+        send_event_on_commit(realm, event, [user_profile.id])
+        user_ids_to_send_event.remove(user_profile.id)
+
+    if updated_status_fields:
+        event = dict(
+            type="user_status",
+            user_id=user_profile.id,
+            **updated_status_fields,
+        )
+        send_event_on_commit(realm, event, user_ids_to_send_event)
+
+
+def try_clear_scheduled_user_status() -> bool:
+    user_statuses = UserStatus.objects.filter(scheduled_end_time__lte=timezone_now()).order_by(
+        "scheduled_end_time"
+    )
+
+    if user_statuses.exists():
+        for user_status in user_statuses:
```

Comment:
`.exists()` wastefully runs a separate database query. (It would be more efficient if we were not also loading the data, but we are.) Do this instead.

```suggestion
    if user_statuses:
        for user_status in user_statuses:
```

Reply:
I guess the `if` condition can be removed altogether, as there is no return value for this function now.